### PR TITLE
[SHC-330] Adding ability to set first and last splits, so that non-alphabetic splits can be easily supported

### PR DESCRIPTION
**Pull request by davidov541**
## What changes were proposed in this pull request?

Adds two properties that allow the first and last split to be set by the user. This allows the use of more complex split preferences, as well as enables scenarios such as numeric-only keys, which were underserved in earlier versions.

## How was this patch tested?

Two unit tests were added. One tests that for alphabetic keys, the rows are dispersed between regions correctly. The other uses numeric keys and sets the first and last splits accordingly. This also checks that the rows have been dispersed appropriately across regions.

### CODE REVIEW

**Comment by davidov541** [comment]
Fixes issue https://github.com/hortonworks-spark/shc/issues/330

**Comment by peter-toth** [comment]
Looks good to me, @dongjoon-hyun could you pleas take a look as well?

**Review by peter-toth — COMMENTED** [review]
Why do we need `.github/workflows/maven.yml` in this PR?

**Comment by davidov541** [comment]
@peter-toth : We don't need maven.yml. I suspect that was added automatically when I was trying to set up a maven repository to help the customer, and didn't realize it was included as part of the PR. It's now been removed.

**Comment by cricket007** [comment]
Might want to add `.github/workflows` to the gitignore until if/when TravisCI is migrated to Github Actions.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 1]
```diff
@@ -250,6 +251,8 @@ object HBaseTableCatalog {
   val tableCoder = "tableCoder"
   // The version number of catalog
   val cVersion = "version"
+  val minRegionSplit = "minregionsplit"
+  val maxRegionSplit = "maxregionsplit"
```
Comment (cricket007) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Not familar with where `parameters` is defined but I would assume these strings are  camelCase

**Comment by davidov541** [comment]
@cricket007 Fixed the internal names. Thanks!

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 2]
```diff
@@ -250,6 +251,8 @@ object HBaseTableCatalog {
   val tableCoder = "tableCoder"
   // The version number of catalog
   val cVersion = "version"
+  val minRegionSplit = "minRegionSplit"
+  val maxRegionSplit = "maxRegionSplit"
```
Comment (joshelser) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: While I can appreciate these names and what you're trying to allow the user to provide, they aren't using HBase nomenclature well.

I think the names `minTableSplitPoint` and `maxTableSplitPoint` would get the point across better what they're doing.

**Review by joshelser — COMMENTED** [review]
Thanks for putting this together, David.

Having read this a few times, the logic on the number of regions which will be created by SHC is a little... obtuse, to say the least. Best as I can see, SHC is doing everything correctly to avoid an error, but table distribution (the split points applied) might be surprising to a user.

They will get, at a minimum, 4 regions because a minimum of three split-points are chosen: the "min", the "max", and one split point lexicographically equidistant between the min and the max (e.g. if `min=A` and `max=Z`, we'd split between `M` and `N`). If the user requests more regions, we will generate more split points between the min and the max.

That's all to say, I think the user documentation could benefit from explaining this somewhere, rather than let the user figure this out via experimentation :). Maybe we can add something to the README as a follow-up?

**Comment by davidov541** [comment]
Thanks, @joshelser. I'm working on a rename of the parameter to match your request. I think that seems reasonable. Not being an HBase expert myself, I wasn't clear on the difference there, but I get it.

As for the README, I agree. In general, the README could use a lot of work. My understanding is that this project is EOL essentially. If so, is it still worth updating the README at this point in time? If the PR will be reviewed and accepted in a reasonable amount of time, I'd be down to update the README at the very least for this portion.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 2 above)
Comment (davidov541) [reply core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Change has been made. Thanks @joshelser!

**Review by peter-toth — APPROVED** [review]
LGTM. Thanks.

@joshelser please review before I merge it.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 2 above)
Comment (peter-toth) [reply core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: @joshelser, these parameters are called now `minRegionSplitPoint` and `maxRegionSplitPoint`. Do you think `minTableSplitPoint` and `maxTableSplitPoint` would be better names?

**Comment by joshelser** [comment]
> please review before I merge it.

Looks like the configuration names changed, but not to what I had suggested: `minRegionSplitPoint` and `maxRegionSplitPoint`. I suggested `minTableSplitPoint` and `maxTableSplitPoint`. We're adding splits to a table which creates more regions. We're not adding split points to Regions. Sorry for being pedantic.

> My understanding is that this project is EOL essentially. If so, is it still worth updating the README at this point in time

Great question that I don't have a good answer for :). Usually I follow a rule of "if I come back to something for a second time, I should fix the documentation because there will be a third time." Ultimately it's up to you. Hopefully we have the right folks with the right permissions to make changes going forward.

**Comment by davidov541** [comment]
@joshelser : Don't worry about it. It's my fault for not paying enough attention. I'll get on that and update the PR in the next week or two.

I'll go ahead and make the change we've discussed to the README and submit a PR to everyone at some point in the next week.

**Review by joshelser — APPROVED** [review]
Thanks for fixing, David!

**Comment by peter-toth** [comment]
Thanks @davidov541 for your work. I've merged it to `master`.

**Pull request merged**